src/cts/src/HTreeBuilder.cpp

```diff
@@ -224,6 +225,456 @@ void HTreeBuilder::initSinkRegion()
+                        ans,
+                        d,
+                        leng,
+                        d / leng);
+        branchPoint.setX(ans.getX());
+        branchPoint.setY(ans.getY());
+      }
+    }
+  }
+}
+
+void HTreeBuilder::legalize()
+{
+  Point<double> topLevelBufferLoc = sinkRegion_.computeCenter();
+  for (int levelIdx = 0; levelIdx < topologyForEachLevel_.size(); ++levelIdx) {
+    LevelTopology& topology = topologyForEachLevel_[levelIdx];
+
+    for (unsigned idx = 0; idx < topology.getBranchingPointSize(); ++idx) {
+      // idx is the buffer id at level levelIdx
+      Point<double>& branchPoint = topology.getBranchingPoint(idx);
+      unsigned parentIdx = topology.getBranchingPointParentIdx(idx);
+
+      Point<double> parentPoint
+          = (levelIdx == 0)
+                ? topLevelBufferLoc
+                : topologyForEachLevel_[levelIdx - 1].getBranchingPoint(
+                    parentIdx);
+
+      const std::vector<Point<double>>& sinks
+          = topology.getBranchSinksLocations(idx);
+      // if ( sinks.empty() ) continue ;
+
+      Point<double>& qt = branchPoint;
+      double leng = topology.getLength();
+
+      // xxx may change topLevelBufferLoc even when it is not inside blockage
+      /*
+        if ( levelIdx==1 && idx==0 ) { // only for the root that is at the
+        top-evel buffer (at LevelTopology& top = topologyForEachLevel_[0];
+        Point<double>& a  = top.getBranchingPoint(0) ; // left child
+        Point<double>& b  = top.getBranchingPoint(1) ; // right child
+        adjustToplevelTopology ( a,b, topLevelBufferLoc );
+        double da = a.computeDist ( topLevelBufferLoc );
+        double db = b.computeDist ( topLevelBufferLoc );
+        logger_->report("yy8 da,db = {},{}", da, db );
+        }
+      */
+
+      int z = wireSegmentUnit_;
+      double x1, y1, x2, y2;
+      odb::dbBlockage* obs = findBlockage(qt, z, x1, y1, x2, y2, db_);
+      if (obs != nullptr) {
+        Point<double> ans(qt);
+        if (levelIdx == 0) {
+          bool moved = moveOnBlockBounardy(parentPoint, ans, x1, y1, x2, y2);
+          if (moved) {
+            // logger_->report("xxx yy8 top {}:{}-->{} ", levelIdx, qt, ans );
+          }
+        } else {
+          if (levelIdx == 1) {
+            leng = qt.computeDist(parentPoint);
+            topology.setLength(leng);
+          }
+          std::vector<Point<double>> points;
+          // find all the possible locations off the blockage
+          findLegalPlacement(parentPoint, leng, x1, y1, x2, y2, points);
+          // std::string name = obs->getName();
+          std::string name = "xxx";
+          // choose the best new location
+          ans = selectBestNewLocation(qt, points, sinks);
+          if (levelIdx == 2) {
+            double moved = ans.computeDist(qt);
+            logger_->report(
+                "xxx y8 top {}:{}-->{}, moved={} ", levelIdx, qt, ans, moved);
+          }
```

Comment:
unhelpful message - looks to be debug code.  Convert to debugPrint or rm

Reply:
Removed